README.md
# Windows Azure Active Directory Authentication Library (ADAL) for Node.js
The ADAL for node.js library makes it easy for node.js applications to authenticate to AAD in order to access AAD protected web resources.  It supports 3 authentication modes shown in the quickstart code below.

## Versions
Current version - 0.2.2  
Minimum recommended version - 0.2.2  
You can find the changes for each version in the [change log](https://github.com/AzureAD/azure-activedirectory-library-for-nodejs/blob/master/changelog.txt).

## Samples and Documentation

[We provide a full suite of sample applications and documentation on GitHub](https://github.com/azure-samples?q=active-directory) to help you get started with learning the Azure Identity system. This includes tutorials for native clients such as Windows, Windows Phone, iOS, OSX, Android, and Linux. We also provide full walkthroughs for authentication flows such as OAuth2, OpenID Connect, Graph API, and other awesome features. 

## Community Help and Support

We leverage [Stack Overflow](http://stackoverflow.com/) to work with the community on supporting Azure Active Directory and its SDKs, including this one! We highly recommend you ask your questions on Stack Overflow (we're all on there!) Also browse existing issues to see if someone has had your question before. 

We recommend you use the "adal" tag so we can see it! Here is the latest Q&A on Stack Overflow for ADAL: [http://stackoverflow.com/questions/tagged/adal](http://stackoverflow.com/questions/tagged/adal)

## Security Reporting

If you find a security issue with our libraries or services please report it to [secure@microsoft.com](mailto:secure@microsoft.com) with as much detail as possible. Your submission may be eligible for a bounty through the [Microsoft Bounty](http://aka.ms/bugbounty) program. Please do not post security issues to GitHub Issues or any other public site. We will contact you shortly upon receiving the information. We encourage you to get notifications of when security incidents occur by visiting [this page](https://technet.microsoft.com/en-us/security/dd252948) and subscribing to Security Advisory Alerts.

## Contributing

All code is licensed under the Apache 2.0 license and we triage actively on GitHub. We enthusiastically welcome contributions and feedback. You can clone the repo and start contributing now. 

## Quick Start
### Installation

``` $ npm install adal-node ```

### Configure the logging

#### Personal Identifiable Information (PII) & Organizational Identifiable Information (OII)

By default, ADAL logging does not capture or log any PII or OII. The library allows app developers to turn this on by configuring the `loggingWithPII` flag in the logging options. By turning on PII or OII, the app takes responsibility for safely handling highly-sensitive data and complying with any regulatory requirements.

```javascript
var logging = require('adal-node').Logging;

//PII or OII logging disabled. Default Logger does not capture any PII or OII.
logging.setLoggingOptions({
  log: function(level, message, error) {
    // provide your own implementation of the log function
  },
  level: logging.LOGGING_LEVEL.VERBOSE, // provide the logging level
  loggingWithPII: false  // Determine if you want to log personal identification information. The default value is false.
});

//PII or OII logging enabled.
logging.setLoggingOptions({
  log: function(level, message, error) {
    // provide your own implementation of the log function
  },
  level: logging.LOGGING_LEVEL.VERBOSE,
  loggingWithPII: true
});
```

### Authorization Code

See the [website sample](https://github.com/MSOpenTech/azure-activedirectory-library-for-nodejs/blob/master/sample/website-sample.js) for a complete bare bones express based web site that makes use of the code below.

```javascript
var AuthenticationContext = require('adal-node').AuthenticationContext;

var clientId = 'yourClientIdHere';
var clientSecret = 'yourAADIssuedClientSecretHere'
var authorityHostUrl = 'https://login.windows.net';
var tenant = 'myTenant';
var authorityUrl = authorityHostUrl + '/' + tenant;
var redirectUri = 'http://localhost:3000/getAToken';
var resource = '00000002-0000-0000-c000-000000000000';
var templateAuthzUrl = 'https://login.windows.net/' + 
                        tenant + 
                        '/oauth2/authorize?response_type=code&client_id=' +
                        clientId + 
                        '&redirect_uri=' + 
                        redirectUri + 
                        '&state=<state>&resource=' + 
                        resource;

function createAuthorizationUrl(state) {
  return templateAuthzUrl.replace('<state>', state);
}

// Clients get redirected here in order to create an OAuth authorize url and redirect them to AAD.
// There they will authenticate and give their consent to allow this app access to
// some resource they own.
app.get('/auth', function(req, res) {
  crypto.randomBytes(48, function(ex, buf) {
    var token = buf.toString('base64').replace(/\//g,'_').replace(/\+/g,'-');

    res.cookie('authstate', token);
    var authorizationUrl = createAuthorizationUrl(token);

    res.redirect(authorizationUrl);
  });
});

// After consent is granted AAD redirects here.  The ADAL library is invoked via the
// AuthenticationContext and retrieves an access token that can be used to access the
// user owned resource.
app.get('/getAToken', function(req, res) {
  if (req.cookies.authstate !== req.query.state) {
    res.send('error: state does not match');
  }

  var authenticationContext = new AuthenticationContext(authorityUrl);

  authenticationContext.acquireTokenWithAuthorizationCode(
    req.query.code,
    redirectUri,
    resource,
    clientId, 
    clientSecret,
    function(err, response) {
      var errorMessage = '';
      if (err) {
        errorMessage = 'error: ' + err.message + '\n';
      }
      errorMessage += 'response: ' + JSON.stringify(response);
      res.send(errorMessage);
    }
  );
});
```

### Server to Server via Client Credentials

See the [client credentials sample](https://github.com/MSOpenTech/azure-activedirectory-library-for-nodejs/blob/master/sample/client-credentials-sample.js).

```javascript
var AuthenticationContext = require('adal-node').AuthenticationContext;

var authorityHostUrl = 'https://login.windows.net';
var tenant = 'myTenant.onmicrosoft.com'; // AAD Tenant name.
var authorityUrl = authorityHostUrl + '/' + tenant;
var applicationId = 'yourApplicationIdHere'; // Application Id of app registered under AAD.
var clientSecret = 'yourAADIssuedClientSecretHere'; // Secret generated for app. Read this environment variable.
var resource = '00000002-0000-0000-c000-000000000000'; // URI that identifies the resource for which the token is valid.

var context = new AuthenticationContext(authorityUrl);

context.acquireTokenWithClientCredentials(resource, applicationId, clientSecret, function(err, tokenResponse) {
  if (err) {
    console.log('well that didn\'t work: ' + err.stack);
  } else {
    console.log(tokenResponse);
  }
});
```

## License
Copyright (c) Microsoft Open Technologies, Inc.  All rights reserved. Licensed under the Apache License, Version 2.0 (the "License"); 

## We Value and Adhere to the Microsoft Open Source Code of Conduct

This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.


changelog.md
Version 0.2.2-beta
------------------
Release Date: 1 February 2020
* Replace 'request` with `axios`

Version 0.2.0
------------------
Release Date: 23 August 2019
* Add support for user provided api version

Version 0.1.28
--------------
Release Date: 26 Feburary 2018
* Added GDPR support per Microsoft policy.

Version 0.1.27
--------------
Release Date: 08 January 2018
* Fix Issue #185 Dependency "xpath.js" version does not have an OSI-approved license

Version 0.1.26
--------------
Release Date: 05 December 2017
* Added .npmignore to avoid publishing unnecessary files

Version 0.1.25
--------------
Release Date: 06 November 2017
* Fixed typing: acquireUserCode returns UserCodeInfo in the callback

Version 0.1.24
--------------
Release Date: 06 November 2017
* Added type definitions to the project. This should help users developing their apps in TypeScript #142, [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/12586).
* Replaced `"node-uuid"` with `"uuid"` #170.

Version 0.1.23
--------------
Release Date: 26 Oct 2017
* Fix Issue #95 - Change npm licenses to license and use spdx syntax 
* Fix Issue #122 - Broken link on README file 
* Fix Issue #155 - remove need for dynamic access to package.json to obtain lib version (util.js)
* Fix Issue #172 - Add `login.microsoftonline.us` endpoint
* Added Windows Graph Sample 

Version 0.1.22
--------------
Release Date: 1 Aug 2016
* Fix Issue #132 - Trim developer provided authority to construct endpoints

Version 0.1.21
--------------
Release Date: 23 Jun 2016
* Policheck and credscan fixes

Version 0.1.20
--------------
Release Date: 17 Jun 2016
* Add support for resource owner grant flow for ADFS

Version 0.1.19
--------------
Release Date: 26 Apr 2016
* Fixed CredScan issue for the checked in pem file
* Policheck fixes
* Updated node-uuid
* Fixed Readme.md
* Inline the adal version during compile time
* Fix issue #71 - client-credential cert-bad-cert failing
* Fix issue #78 - Express dependency in package.json doesn't match syntax used in website-sample.js
* Fix issue #80 - Unreachable ADFS server results in misleading error message

Version 0.1.18
--------------
Release Date: 5 Feb 2016
* Add oid in IDTokenMap to expose for returned user info. 

Version 0.1.17
--------------
Release Date: 8 Oct 2015
* Add support for cross tenant refresh token

Version 0.1.16
--------------
Release Date: 3 Sep 2015
* Add support for device profile 

Version 0.1.15
--------------
Release Date: 4 Aug 2015
* Fix issue #68 - add support for wstrust 2005 endpoint. 
* Fix issue #62 - escape xml chars in password before creating a xml request for wstrust. 

Version 0.1.14
--------------
Release Date: 5 May 2015
* Add timestamp to the log entries.

Version 0.1.13
--------------
Release Date: 1 May 2015
* Scrub security sensitive data in WS-Trust exchange from log messages.
* Update the version of the jws dependency to the latest release.

Version 0.1.12
--------------
Release Date: 19 February 2015
* Add login.microsoftonline.com to the static list of known authorities.
* Bug fixes. (#36)

Version 0.1.11
--------------
Release Date: 16 December 2014
* Added support for certificate authentication for confidential clients.

Version 0.1.10
--------------
Release Date: 24 November 2014
* 0.1.9 published version was corrupt.

Version 0.1.9
--------------
Release Date: 24 November 2014
* Fixed issue #22 - ADAL needs a method that allows confidential clients
to acquire a new token via a refresh token.The `acquireTokenWithRefreshToken` function was updated to take an additional `clientSecret` argument. Passing this new argument is optional and the change should not breakapps using this function before the additional argument was added.

Version 0.1.8
--------------
Release Date: 29 October 2014
* Update version of xpath.js dependency to 1.0.5
* Fix a bug in the regular expression used to parse a 401 challenge.

Version 0.1.7
--------------
Release Date: 2 September 2014
* Add caching support to AuthenticationContext.acquireTokenWithClientCredentials


lib/constants.js
/*
 * @copyright
 * Copyright © Microsoft Open Technologies, Inc.
 *
 * All Rights Reserved
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http: *www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
 * OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION
 * ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A
 * PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
 *
 * See the Apache License, Version 2.0 for the specific language
 * governing permissions and limitations under the License.
 */
'use strict';

var Constants = {
  OAuth2 : {
      Parameters : {
        GRANT_TYPE : 'grant_type',
        CLIENT_ASSERTION : 'client_assertion',
        CLIENT_ASSERTION_TYPE : 'client_assertion_type',
        CLIENT_ID : 'client_id',
        CLIENT_SECRET : 'client_secret',
        REDIRECT_URI : 'redirect_uri',
        RESOURCE : 'resource',
        CODE : 'code',
        SCOPE : 'scope',
        ASSERTION : 'assertion',
        AAD_API_VERSION : 'api-version',
        USERNAME : 'username',
        PASSWORD : 'password',
        REFRESH_TOKEN : 'refresh_token',
        LANGUAGE : 'mkt',
        DEVICE_CODE : 'device_code',
        REQUESTED_TOKEN_USE : 'requested_token_use'
      },

      GrantType : {
        AUTHORIZATION_CODE : 'authorization_code',
        REFRESH_TOKEN : 'refresh_token',
        CLIENT_CREDENTIALS : 'client_credentials',
        JWT_BEARER : 'urn:ietf:params:oauth:grant-type:jwt-bearer',
        PASSWORD : 'password',
        SAML1 : 'urn:ietf:params:oauth:grant-type:saml1_1-bearer',
        SAML2 : 'urn:ietf:params:oauth:grant-type:saml2-bearer',
        DEVICE_CODE: 'device_code'
      },

      ClientAssertionType : {
        JWT_BEARER: 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer'
      },

      ResponseParameters : {
        CODE : 'code',
        TOKEN_TYPE : 'token_type',
        ACCESS_TOKEN : 'access_token',
        ID_TOKEN : 'id_token',
        REFRESH_TOKEN : 'refresh_token',
        CREATED_ON : 'created_on',
        EXPIRES_ON : 'expires_on',
        EXPIRES_IN : 'expires_in',
        RESOURCE : 'resource',
        ERROR : 'error',
        ERROR_DESCRIPTION : 'error_description'
      },

      DeviceCodeResponseParameters : {
         USER_CODE : 'user_code', 
         DEVICE_CODE : 'device_code', 
         VERIFICATION_URL : 'verification_url',
         EXPIRES_IN : 'expires_in', 
         INTERVAL: 'interval', 
         MESSAGE: 'message', 
         ERROR: 'error', 
         ERROR_DESCRIPTION: 'error_description'
      },

      Scope : {
        OPENID : 'openid'
      },

      RequestedTokenUse : {
        ON_BEHALF_OF : 'on_behalf_of'
      },

      IdTokenMap : {
        'tid' : 'tenantId',
        'given_name' : 'givenName',
        'family_name' : 'familyName',
        'idp' : 'identityProvider',
        'oid': 'oid'
      }
    },

    TokenResponseFields : {
      TOKEN_TYPE : 'tokenType',
      ACCESS_TOKEN : 'accessToken',
      REFRESH_TOKEN : 'refreshToken',
      CREATED_ON : 'createdOn',
      EXPIRES_ON : 'expiresOn',
      EXPIRES_IN : 'expiresIn',
      RESOURCE : 'resource',
      USER_ID : 'userId',
      ERROR : 'error',
      ERROR_DESCRIPTION : 'errorDescription'
    },

    UserCodeResponseFields : {
        USER_CODE : 'userCode', 
        DEVICE_CODE: 'deviceCode', 
        VERIFICATION_URL: 'verificationUrl',
        EXPIRES_IN: 'expiresIn', 
        INTERVAL: 'interval', 
        MESSAGE: 'message', 
        ERROR: 'error', 
        ERROR_DESCRIPTION: 'errorDescription'
    },

    IdTokenFields : {
      USER_ID : 'userId',
      IS_USER_ID_DISPLAYABLE : 'isUserIdDisplayable',
      TENANT_ID : 'tenantId',
      GIVE_NAME : 'givenName',
      FAMILY_NAME : 'familyName',
      IDENTITY_PROVIDER : 'identityProvider'
    },

    Misc : {
      MAX_DATE : 0xffffffff,
      CLOCK_BUFFER : 5 // In minutes.
    },

    Jwt : {
      SELF_SIGNED_JWT_LIFETIME : 10, // 10 mins in mins
      AUDIENCE : 'aud',
      ISSUER : 'iss',
      SUBJECT : 'sub',
      NOT_BEFORE : 'nbf',
      EXPIRES_ON : 'exp',
      JWT_ID : 'jti'
    },

    AADConstants : {
      WORLD_WIDE_AUTHORITY : 'login.windows.net',
      WELL_KNOWN_AUTHORITY_HOSTS : ['login.windows.net', 'login.microsoftonline.com', 'login.chinacloudapi.cn', 'login-us.microsoftonline.com', 'login.microsoftonline.de'],
      INSTANCE_DISCOVERY_ENDPOINT_TEMPLATE : 'https://{authorize_host}/common/discovery/instance?authorization_endpoint={authorize_endpoint}&api-version=1.0',
      AUTHORIZE_ENDPOINT_PATH : '/oauth2/authorize',
      TOKEN_ENDPOINT_PATH : '/oauth2/token', 
      DEVICE_ENDPOINT_PATH : '/oauth2/devicecode'
    },

    UserRealm : {
      FederationProtocolType : {
        WSFederation : 'wstrust',
        SAML2 : 'saml20',
        Unknown : 'unknown'
      },

      AccountType : {
        Federated : 'federated',
        Managed : 'managed',
        Unknown : 'unknown'
      }
    },

    Saml : {
      TokenTypeV1 : 'urn:oasis:names:tc:SAML:1.0:assertion',
      TokenTypeV2 : 'urn:oasis:names:tc:SAML:2.0:assertion'
    },

    XmlNamespaces : {
      wsdl   : 'http://schemas.xmlsoap.org/wsdl/',
      sp     : 'http://docs.oasis-open.org/ws-sx/ws-securitypolicy/200702',
      sp2005 : 'http://schemas.xmlsoap.org/ws/2005/07/securitypolicy',
      wsu    : 'http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd',
      wsa10  : 'http://www.w3.org/2005/08/addressing',
      http   : 'http://schemas.microsoft.com/ws/06/2004/policy/http',
      soap12 : 'http://schemas.xmlsoap.org/wsdl/soap12/',
      wsp    : 'http://schemas.xmlsoap.org/ws/2004/09/policy',
      s      : 'http://www.w3.org/2003/05/soap-envelope',
      wsa    : 'http://www.w3.org/2005/08/addressing',
      wst    : 'http://docs.oasis-open.org/ws-sx/ws-trust/200512', 
      t      : 'http://schemas.xmlsoap.org/ws/2005/02/trust'
    },

    Cache : {
      HASH_ALGORITHM : 'sha256'
    },

    HttpError : {
      UNAUTHORIZED : 401
    },

    AdalIdParameters : {
      SKU : 'x-client-SKU',
      VERSION : 'x-client-Ver',
      OS : 'x-client-OS',
      CPU : 'x-client-CPU',
      NODE_SKU : 'Node'
    },

    WSTrustVersion : {
       UNDEFINED : 'undefined', 
       WSTRUST13 : 'wstrust13', 
       WSTRUST2005 : 'wstrust2005'
    } 
  };

module.exports = Constants;
